### PR TITLE
Implement Aggregate Relationships

### PR DESCRIPTION
**Pull request by tnightengale**
https://github.com/tnightengale/dbt-activity-schema/issues/9

### CODE REVIEW

File: macros/relationships/append_only/aggregate_in_before.sql [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+(
+    {{ stream(i) }}.{{- columns.ts }} < {{ stream() }}.{{- columns.ts }}
+)
+{% endmacro %}
+
+{% macro aggregate_in_before_aggregation_func() %}
+count(distinct {{ caller() }})
+{% endmacro %}
+
+{% macro aggregate_in_before() %}
```
Comment (tnightengale) [inline macros/relationships/append_only/aggregate_in_before.sql:15]: I was thinking we could actually allow users to pass their own aggregation macros here:

```suggestion
{% macro aggregate_in_before(agg_func=dbt_activity_schema.aggregate_in_before_aggregation_func) %}
```

File: macros/relationships/append_only/aggregate_in_before.sql [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+{% do return(namespace(
+    name="aggregate_in_before",
+    aggregation_func=dbt_activity_schema.aggregate_in_before_aggregation_func,
```
Comment (tnightengale) [inline macros/relationships/append_only/aggregate_in_before.sql:19]:
```suggestion
    aggregation_func=agg_func,
```

File: macros/relationships/append_only/aggregate_in_before.sql [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+{% set stream = dbt_activity_schema.generate_stream_alias %}
+{% set columns = dbt_activity_schema.columns() %}
+
+(
+    {{ stream(i) }}.{{- columns.ts }} < {{ stream() }}.{{- columns.ts }}
+)
+{% endmacro %}
+
+{% macro aggregate_in_before_aggregation_func() %}
+count(distinct {{ caller() }})
```
Comment (tnightengale) [inline macros/relationships/append_only/aggregate_in_before.sql:12]: Caller functionally in jinja allows the macro to act as a decorator. See here for details: https://jinja.palletsprojects.com/en/3.1.x/templates/#call

File: macros/utils/aggregations/max.sql [hunk 4]
```diff
@@ -0,0 +1,3 @@
+{% macro max() %}
+max({{ caller() }})
```
Comment (bcodell) [inline macros/utils/aggregations/max.sql:2]: Looking for context - what exactly does `caller()` do?

Comment (bcodell) [reply macros/utils/aggregations/max.sql:2]: nvm - missed the comment above 🤦

**Comment by tnightengale** [comment]
@bcodell I am going to break out the `caller()` aggregation implementation into a separate PR.

**Review by bcodell — APPROVED** [review]
Looks good! General note that adding descriptions in yaml where tests are specified would be helpful for faster understanding of the purpose of each test during PR review and for future reference to know what is/isn't covered, so that additional tests can be added when needed. Not a blocker though so approving 👍

**Comment by tnightengale** [comment]
@bcodell I think all your comments were on outdated commits; perhaps the page you started the review from was not refreshed?

Most of the comments were actually handled in the way you suggested in the latest commits. To this point:
> Yeah this feels like the right approach. Does this argument always need to be a macro because of the way call blocks end up getting used? Or could someone pass a plain text string like 'sum' or 'list_agg'?

Yes, the always need to use the `{{ caller() }}` builtin, strings will not work.

Also:
> I like this better than the string-based approach, but if there's a possible future where more aggregation functions will be supported, then I think a better interface would be to create a registry of sorts to access all aggregations - similar to the columns() interface - so something like dbt_activity_schema.aggregations().min in this example. It's more verbose, but it feels more organized than having the different aggregation methods floating around at the top level of the package.

Yes, I agree we could namespace these more nicely 👍

**Comment by bcodell** [comment]
Ah whoops - I added my approving PR through GitHub mobile and I probably misused the UI 🤷‍♂️ 

Nice work!

**Pull request merged**